### ISSUE TITLE
Question about scalars

### ISSUE THREAD

**Issue by sungam3r** [graphql/graphql-spec#826]
See discussion in https://github.com/graphql-dotnet/graphql-dotnet/pull/2357#discussion_r588738887

The spec says 
![изображение](https://user-images.githubusercontent.com/21261007/110234641-a2bd1c80-7f3c-11eb-8f67-6fd361bac466.png)
but technically, it is very easy to make that the scalar will be returned as an object (at least fro GraphQL.NET). The question is whether it is a permissible expansion of behavior or completely unacceptable.

**Comment by benjie** [comment]
The value that is at the tip of the GraphQL tree is a leaf, even if it's an object returned from a custom JSON scalar rather than just a `String`. The word "primitive" here generally implies that it's a simple type, but really it's just a "building block" type - GraphQL sees it as an "atom", indivisible.

That said; the wording here is potentialyl ambiguous because it says the leaves on the trees are GraphQL scalars, but they could also be enums, which in some parts of the spec are treated as within the scope of the term "scalars" and in some parts are not.

The concept of a JSON scalar has been discussed a few times at the GraphQL Spec WG and it's seen as allowable (though it's an "escape hatch" and should be avoided where possible). It has caused us quite a few issues (it's restricted how we can grow certain parts of the spec) but it's something that is in widespread use so we have to continue to support.

**Comment by andimarek** [comment]
To add a bit on that:

the value returned by custom scalar for the result coercion (in GraphQL.js this function is called `serialize`) can by anything that the Programming language/environment allows. For example it can be a `Money` object. 

GraphQL in general doesn't not limit at all this aspect: it is an implementation detail of the GraphQL implementation. 

To phrase it a bit different: The result of an GraphQL execution is tree, where Scalars or Enums are the leaves from a type perspective. But what the actual values of these types are are runtime can't be predicted or described by the spec. 

Normally In practical terms a complex value returned by a Scalar result coercion probably needs to be serialized (warning: horrible naming, this a transport serialize which is a separate step from the result coercion) into a JSON representation. 

Have a look at https://www.graphql-scalars.com/scalars-guide and https://www.graphql.de/blog/scalars-in-depth/ for more details.

**Comment by sungam3r** [comment]
@benjie @andimarek To summarize:
1. It is absolutely normal that scalars can eventually be represented by objects after serialization in JSON.
2. The specification explicitly does not speak about it.

Right? If so, doesn't you think that the specification needs to be supplemented in such a way as to eliminate the possibility of different interpretations?

**Comment by sungam3r** [comment]
> Have a look at https://www.graphql-scalars.com/scalars-guide 

I read it and even made two PRs last summer.

**Comment by benjie** [comment]
> If so, doesn't you think that the specification needs to be supplemented in such a way as to eliminate the possibility of different interpretations?

Yes, the spec should always be clarified if it's likely there will be multiple interpretations - it's job is to be precise. I suggest you make the relevant edits and add the topic to a GraphQL Working Group :+1:

**Comment by martinbonnin** [comment]
This has been a while. For the record, @benjie just added the definition of a [leaf field](https://github.com/graphql/graphql-spec/pull/1212/changes). @sungam3r please let us know how that sounds to you. Feel free to suggest any correction you'd find useful!

**Comment by martinbonnin** [comment]
I'm going to close this one as stale. Scalars my be serialized as JSON objects, this is perfectly valid. Leaf fields are fields which can be of scalar or enum type.

**Comment by benjie** [comment]
(Or lists thereof. The named type will be a scalar or enum type.)
